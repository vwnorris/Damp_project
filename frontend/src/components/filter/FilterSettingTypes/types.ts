import { IFilterData } from "../../../types";


export interface IFilterSettingTypeProp {
    setFilterData: (data: IFilterData) => void;
}