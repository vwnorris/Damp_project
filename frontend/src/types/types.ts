export interface Comment {
    name: string;
    rating: number;
    comment: string;
}

export interface Game {
    name: string;
    genre: string[];
    release_date: string;
    all_reviews: string;
    imagePath: string;
    publisher: string[];
    developer: string;
    appId: number;
    game_description: string;
    achievements: number;
    languages: string[];
    price: number;
    description: string;
    comments: Comment[];
    releaseDate: Date;
}

export enum SortType {
    NAME = "NAME",
    RELEASEDATE = "RELEASEDATE",
    PRICE = "PRICE",
    NONE = "NONE"
}

export interface IFilterData {
    type: string;
    data: number | string[] | number[] | (Date | null)[];
    visualData: string;
}

export interface IFilter {
    name: string;
    color: string;
    description: string;
    data?: IFilterData;
}