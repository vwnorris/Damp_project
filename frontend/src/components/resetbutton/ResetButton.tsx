import { Button } from "@mui/material";
import { observer } from "mobx-react";
import React, { useEffect, useState } from "react";
import { useStores } from "../../hooks";

export const ResetButton = observer(() => {
  const [showButton, setShowButton] = useState<boolean>(false);
  const { store } = useStores();

  useEffect(() => {
    setShowButton(store.enableResetButton);
  }, [store.enableResetButton]);

  const handleReset = () => {
    store.resetStores();
  };

  return (
    <div>
      {showButton && (
        <Button
          data-testid={"reset-button"}
          variant="outlined"
          color="error"
          sx={{
            width: {
              sx: 1,
              md: 0.1,
            },
            borderRadius: 1,
          }}
          onClick={handleReset}
        >
          Reset
        </Button>
      )}
    </div>
  );
});
