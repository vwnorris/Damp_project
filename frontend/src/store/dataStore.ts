import { action, computed, IObservableArray, makeObservable, observable, runInAction } from "mobx";
import { Game, SortType } from "../types";
import { LOAD_GAMES } from "../graphQL";
import { client } from "../util";
import { ApolloClient, ApolloError, NormalizedCacheObject, ObservableQuery } from "@apollo/client";
import { RootStore } from "./rootStore";

interface ISort {
    ascending: boolean;
    type: SortType;
}

/**
 * Store handling data fetching and querying the database
 */
export class DataStore {
    data: IObservableArray<Game>;
    pageNumber: number;
    searchString: string = "";
    loading: boolean;
    allFound: boolean = false;
    error: ApolloError | undefined;
    sort: ISort;
    rootStore: RootStore
    query: ObservableQuery;

    constructor(rootStore: RootStore, inClient: ApolloClient<NormalizedCacheObject> = client) {
        this.query = inClient.watchQuery({query: LOAD_GAMES});
        this.pageNumber = 0
        this.loading = false;
        this.error = undefined;
        this.data = observable.array<Game>([]);
        this.sort = {
            ascending: true,
            type: SortType.NONE
        };
        this.rootStore = rootStore;
        makeObservable(this, {
            data: observable,
            loading: observable,
            error: observable,
            allFound: observable,
            searchString: observable,
            sort: observable,
            getMoreData: action,
            reloadData: action,
            setSort: action,
            setSearchString: action,
            resetStore: action,
            resetable: computed
        });
        runInAction(() => {
            this.getMoreData();
        });
    }

    get resetable(): boolean {
        if (this.pageNumber > 1 || this.data.length > 15 || this.sort.type !== SortType.NONE) return true
        return false;
    }

    async reloadData() {
        runInAction(() => {
            this.pageNumber = 0;
            this.data.clear();
            this.allFound = false;
            this.getMoreData();
        });
    }

    setSearchString(s: string) {
        this.searchString = s;
    }

    async getMoreData() {
        const variables = {
            page: this.pageNumber,
            filter: this.rootStore.filterStore.queryFilter,
            sort: this.sort,
            search: this.searchString
        };
        this.error = undefined;
        this.loading = true;
        this.query.refetch(
            variables
        ).then(
            action("getSucces", ({data}) => {
                this.data.push(...data.games);
                this.loading = false;
                this.pageNumber++;
                if (data.games.length < 15) {
                    this.allFound = true;
                }
            })
        ).catch(action("getError", (error: ApolloError) => {
            this.error = error;
        }))
    }

    setSort(type: SortType, ascending: boolean = true) {
        this.sort.ascending = ascending;
        this.sort.type = type;
    }

    resetStore() {
        this.pageNumber = 0
        this.loading = false;
        this.error = undefined;
        this.data.clear();
        this.sort = {
            ascending: true,
            type: SortType.NONE
        };
    }

}