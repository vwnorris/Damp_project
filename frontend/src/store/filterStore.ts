import { observable, action, makeObservable, runInAction, computed } from 'mobx';
import { IFilter, IFilterData } from '../types';
import { RootStore } from './rootStore';

interface ReturnData {
    genre?: string[];
    price?: number[];
    releasedate?: (Date | null)[];
    achivements?: number;
    tags?: string[];
}

/**
 * Store that handles filters and lets the data store get a query to send to the database
 */
export class FilterStore {
    // used sets here just because we want only unique values
    activeFilters: Set<IFilter> = new Set<IFilter>();
    allFilters: Set<IFilter> = new Set<IFilter>([
        {name: "Genre", color: "#F2994A", description: "The genre of the game"}, 
        {name: "Price", color: "#F2C94C", description: "The price of the game"}, 
        {name: 'ReleaseDate', color: '#BB6BD9', description: "The release date of the game"}, 
        {name: "Achievements", color: "#2D9CDB", description: "The number of achievements of the game"}, 
        {name: "Tags", color: "#EB5757", description: "The tags of the game"}]);
    queryFilter: ReturnData = observable.object<ReturnData>({});
    rootStore: RootStore;
    selectedFilter: IFilter | undefined;
    showModalInfo: boolean = false;

    constructor(rootStore: RootStore) {
        this.rootStore = rootStore;
        makeObservable(this, {
            activeFilters: observable,
            allFilters: observable,
            queryFilter: observable,
            addFilter: action,
            removeFilter: action,
            resetStore: action,
            resetable: computed
        });
    }

    get resetable(): boolean {
        if (Array.from(this.activeFilters).length > 0) return true;
        return false;
    }

    /**
     * Adds a filter to the query when a filter is added such that the data store can use it to make a query with the new parameters
     */
    private addFilterQuery (filter: IFilter) {
        switch (filter.name) {
            case "Genre":
                this.queryFilter.genre = (filter.data!.data as string[]);
                break;
            case "Price":
                this.queryFilter.price = (filter.data!.data as number[]);
                break;
            case "ReleaseDate":
                this.queryFilter.releasedate = (filter.data!.data as (Date | null)[]);
                break;
            case "Achievements":
                this.queryFilter.achivements = (filter.data!.data as number);
                break;
            case "Tags":
                this.queryFilter.tags = (filter.data!.data as string[]);
                break;
        }
    }

    private removeFilterQuery (filter: IFilter) {
        type objectKey = keyof ReturnData;
        delete this.queryFilter[filter.name.toLowerCase() as objectKey];
    }

    private getFromSet(set: Set<IFilter>, name: string): IFilter {
        let elem: IFilter;
        for(let a of Array.from(set)) {
            if (a.name === name) {
                elem = a;
                break;
            }
        }
        set.delete(elem!);
        return elem!;
    }

    addFilter(filter: string, data: IFilterData) {
        const f = this.getFromSet(this.allFilters, filter);
        f.data = data;
        this.addFilterQuery(f);
        this.activeFilters.add(f);
        runInAction(() => {
            this.rootStore.dataStore.reloadData();
        });
    }

    removeFilter(filter: string) {
        const f = this.getFromSet(this.activeFilters, filter);
        this.removeFilterQuery(f);
        this.allFilters.add(f);
        runInAction(() => {
            this.rootStore.dataStore.reloadData();
        });
    }

    selectFilter(filter: IFilter) {
        this.selectedFilter = filter;
        this.showModalInfo = true;
    }

    unSelectFilter() {
        runInAction(() => {
            this.showModalInfo = false;
            this.selectedFilter = undefined;
        });
    }

    resetStore() {
        Array.from(this.activeFilters).forEach(filter => {
            const f = this.getFromSet(this.activeFilters, filter.name);
            this.removeFilterQuery(f);
            this.allFilters.add(f);
        });
    }
}