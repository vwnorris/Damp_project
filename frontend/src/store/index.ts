import { createContext } from "react";
import { RootStore } from "./rootStore";

export const store = new RootStore()

export const defaultContext = createContext({
    store: store
});
