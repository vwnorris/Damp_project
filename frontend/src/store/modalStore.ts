import { ApolloClient, NormalizedCacheObject, ObservableQuery } from "@apollo/client";
import { action, computed, makeObservable, observable, runInAction } from "mobx";
import { GET_SINGLE_GAME } from "../graphQL";
import { Game } from "../types";
import { client } from "../util";
import { RootStore } from "./rootStore";

/**
 * Store handling the modal and data fetching for it
 */
export class ModalStore {
    selectedGame: number = -1;
    game: Game | undefined;
    showModal: boolean = false;
    updating: boolean = false;
    rootStore: RootStore;
    query: ObservableQuery;

    constructor(rootStore: RootStore, inClient: ApolloClient<NormalizedCacheObject> = client) {
        this.rootStore = rootStore;
        this.query = inClient.watchQuery({
            query: GET_SINGLE_GAME
        });
        makeObservable(this, {
            game: observable,
            showModal: observable,
            updating: observable,
            selectGame: action,
            unSelectGame: action,
            updateGameData: action,
            resetable: computed
        });
    }

    get resetable(): boolean {
        if (this.selectedGame > -1 || this.game) return true;
        return false;
    }

    selectGame(appId: number) {
        this.query.refetch({
            appId: appId
        }).then(action("getSuccess", ({data}) => {
            this.game = data.game as Game;
            this.showModal = true;
            this.selectedGame = appId;
        }))
    }

    updateGameData() {
        if (this.selectedGame === -1) return;
        this.updating = true;
        this.query.refetch({
            appId: this.selectedGame
        }).then(action("getSuccess", ({ data }) => {
            this.game = data.game as Game;
            this.updating = false;
        }));
    }

    unSelectGame() {
        runInAction(() => {
            this.showModal = false;
            this.game = undefined;
            this.selectedGame = -1;
        })
    }

    resetStore() {
        this.selectedGame = -1;
        this.game = undefined;
        this.showModal = false;
        this.updating = false;
    }
}