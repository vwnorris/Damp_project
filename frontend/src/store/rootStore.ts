import { ApolloClient, NormalizedCacheObject } from "@apollo/client";
import { action, computed, makeObservable } from "mobx";
import { client } from "../util";
import { DataStore } from "./dataStore";
import { FilterStore } from "./filterStore";
import { ModalStore } from "./modalStore";

/**
 * This class is the backbone of our store system and lets the other stores communicate.
 * It is created with mobx and has our *database* as the default client
 */
export class RootStore {
    filterStore: FilterStore;
    dataStore: DataStore;
    modalStore: ModalStore;

    constructor(inClient: ApolloClient<NormalizedCacheObject> = client) {
        this.filterStore = new FilterStore(this);
        this.dataStore = new DataStore(this, inClient);
        this.modalStore = new ModalStore(this, inClient);
        makeObservable(this, {
            enableResetButton: computed,
            resetStores: action
        });     
    }

    /** Function to see if the reset button should come or not */ 
    get enableResetButton(): boolean {
        if (this.dataStore.resetable || this.modalStore.resetable || this.filterStore.resetable) return true;
        return false;
    }

    /** Function to reset all the stores */ 
    public resetStores(reloadData: boolean = true) {
        this.filterStore.resetStore();
        this.dataStore.resetStore();
        this.modalStore.resetStore();
        if (reloadData) this.dataStore.reloadData();
    }
}