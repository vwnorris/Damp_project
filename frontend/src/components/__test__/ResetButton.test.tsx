import { ResetButton } from "../resetbutton";
import { store } from "../../store"
import { cleanup, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { act } from "react-dom/test-utils";
import renderer from "react-test-renderer";
import { SortType } from "../../types";

beforeEach(() => {
    store.resetStores()
})

afterEach(cleanup)

describe("test header component", () => {
  it("renders with correct header", () => {
    store.dataStore.setSort(SortType.NAME);
    render(<ResetButton />);

    act(() => {
      screen.getByTestId("reset-button").click();
    });
  });
});

describe("header snapshot tests", () => {
  it("basic snapshot test", () => {
    const tree = renderer.create(<ResetButton />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
