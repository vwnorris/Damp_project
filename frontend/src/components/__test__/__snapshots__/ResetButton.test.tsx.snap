// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header snapshot tests basic snapshot test 1`] = `<div />`;
